Add new helper class to wrap RetryHandshake with task support

joseph-wakeling-sociomantic opened the pull request:
The new DhtHandshake class is designed to make it easy for applications to implement support for a partial DHT handshake, e.g.:

```D
auto handshake = new DhtHandshake(theScheduler.epoll, dht_client, 10);

// block until at least one node is connected
theScheduler.await(handshake.oneNodeConnected());

// wait until either all nodes have connected, or 60 seconds
// have passed, whichever comes sooner
theScheduler.awaitOrTimeout(handshake.allNodesConnected(), 60_000_000);
```

The underlying `RetryHandshake` instance should ensure that even if the timeout is hit in the above example, the application will keep trying to contact any missing nodes until it succeeds.

Simple logging has been built into the class in order to make it easy to monitor handshake progress.

Fixes https://github.com/sociomantic-tsunami/dhtproto/issues/126.

joseph-wakeling-sociomantic committed:
Add new helper class to wrap RetryHandshake with task support

The new DhtHandshake class is designed to make it easy for applications
to implement support for a partial DHT handshake, e.g.:

```D
auto retry_delay_seconds = 3
auto handshake = new DhtHandshake(dht_client, retry_delay_seconds);

// block until at least one node is connected
theScheduler.await(handshake.oneNodeConnected());

// wait until either all nodes have connected, or 60 seconds
// have passed, whichever comes sooner
theScheduler.awaitOrTimeout(handshake.allNodesConnected(), 60_000_000);
```

The underlying `RetryHandshake` instance should ensure that even if the
timeout is hit in the above example, the application will keep trying to
contact any missing nodes until it succeeds.

Simple logging has been built into the class in order to make it easy to
monitor handshake progress.

Note that the unittest usage example has been written for compatibility
with the minimum version of ocean supported by dhtproto 13.x.x, which
does not provide the `awaitOrTimeout` method.  This can be updated when
the feature is merged into dhtproto 14.x.x.

By contrast usage examples in release notes and in this commit message
provide the recommended `awaitOrTimeout` usage to support partial DHT
handshakes.

Fixes https://github.com/sociomantic-tsunami/dhtproto/issues/126.

joseph-wakeling-sociomantic commented:
@mihails-strasuns-sociomantic does this fit what you had in mind? :-)

mihails-strasuns-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+
+        protected override void run ()
+        {
+            if (!this.all_dht_nodes_connected) this.suspend();
```
Can also just use  `TaskEvent` here

joseph-wakeling-sociomantic replied:
So basically `run` just contains `TaskEvent.wait();`, and the soon-to-be-renamed `dg` just contains any wanted logging plus `TaskEvent.trigger();` ... ?

mihails-strasuns-sociomantic replied:
Yup

mihails-strasuns-sociomantic replied:
In that sense it just occured  to me that whole handshake helper could be a task event, and not actual task - but it doesn't matter much and this way probably is more generic.

joseph-wakeling-sociomantic replied:
By whole handshake helper you mean the `DhtHandshake` class?  It's not a task itself, it just exposes a couple.

If I understand you right, you mean we could just use a couple of `TaskEvent` instances under the hood and expose methods that would cause the calling task context to invoke their `wait()` ... ?

FWIW I agree that it's nicer/more flexible to expose task instances.  But I don't mind drafting an alternative design (in a separate PR) for comparison.

joseph-wakeling-sociomantic replied:
Probably it could be something like:

```D
class DhtHandshake
{
    private RetryHandshake retry_handshake;

    private TaskEvent one_node_connected;

    private TaskEvent all_nodes_connected;

    public bool start (DhtClient dht, uint timeout, size_t wait_time)
    {
        verify(this.retry_handshake is null);

        void oneNodeConnectedDg (NodeItem item)
        {
            this.one_node_connected.trigger();
        }

        bool timed_out = true;

        void allNodesConnectedDg ()
        {
            timed_out = false;
            this.all_nodes_connected.trigger();
        }

        this.retry_handshake =
            new RetryHandshake(theScheduler.epoll, dht,
                    &allNodesConnectedDg, &this.oneNodeConnectedDg);

        this.one_node_connected.wait();

        if (timeout > 0)
        {
            // launch some timer event that will also call
            // this.all_nodes_connected.trigger()
        }

        this.all_nodes_connected.wait();

        return timed_out;
    }
}
```

... no ... ?

mihails-strasuns-sociomantic replied:
Almost what I had in mind - but you don't need to call `this.one_node_connected.wait();` there, it will be called from host ask (same for `this.all_nodes_connected.wait();`). Benefit with this approach is that you don't need to use 2 worker fibers for one/all cases, it will only suspend one host task - also allocates slightly less memory. Drawback is that it seems slightly more convoluted.

joseph-wakeling-sociomantic replied:
Hmm, are you expecting that the `TaskEvent` fields will be public?  I don't see any other way to viably do what you have in mind given that they are struct instances.  And my inclination was to hide the `trigger` method at least.

FWIW I do think that on balance the existing version using tasks, while a little less efficient (in LoC as well as memory) offers a slightly more intuitive UI.  But do you think it would be worth using `TaskEvent` under the hood of the `{One,All}Node{s}Connected` tasks, in any case?

mihails-strasuns-sociomantic replied:
> Hmm, are you expecting that the TaskEvent fields will be public? I 

You only need to expose public methods like `wait_one_node()` which calls event internally, actual event struct instance is of no other use anyway.

mihails-strasuns-sociomantic replied:
> But do you think it would be worth using TaskEvent under the hood 

Definitely, it easier than to manually look where you have to add `if (suspended)` checks

joseph-wakeling-sociomantic replied:
> You only need to expose public methods like wait_one_node() which calls event internally, actual event struct instance is of no other use anyway.

That sounds like it'll make the timeout less intuitive, though.  There's something really nice about it being in the hands of the user to say: `await(handshake.oneNode())`, `awaitOrTimeout(handshake.allNodes(), ...)`.  So, I think I'll keep the existing public API, but use `TaskEvent` under the hood of the tasks exposed.  SGTY?

mihails-strasuns-sociomantic replied:
Yeah, as I have mentioned before, I am simply calling out loud available options. Current API looks reasonable.

mihails-strasuns-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+            if (!this.at_least_one_node_connected)
+            {
+                this.at_least_one_node_connected = true;
+                if (this.suspended()) this.resume();
```
ditto `taskEvent`

mihails-strasuns-sociomantic reviewed (approved):
Yup, pretty much how I have expected it too look!

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+
+        ***********************************************************************/
+
+        private void dg ()
```
Would be a bit clearer to call this `allNodesConnected`.

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+
+        ***********************************************************************/
+
+        private void dg (NodeItem item)
```
Would be a bit clearer to call this `oneNodeConnected`.

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+
+    ***************************************************************************/
+
+    private Logger log;
```
Optional style point: we tend to prefer `///` comments for everything expect functions, these days.

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+        Constructor
+
+        Params:
+            epoll = epoll select dispatcher via which to schedule handshake
```
If this whole thing is based on tasks, is there any reason to accept an epoll selector here? You could just use `theScheduler.epoll`, internally.

joseph-wakeling-sociomantic replied:
I didn't want to hardcode an unnecessary dependency on `theScheduler`, on the grounds that the user might have a different setup (e.g. custom `Scheduler` instance instead of `theScheduler`).  But that might be over-engineering.

mihails-strasuns-sociomantic replied:
> e.g. custom Scheduler instance instead of theScheduler

This is not supported by ocean

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,211 @@
+    {
+        return this.one_node_connected;
+    }
+}
```
Please add a usage example (like what you included in the commit message).

joseph-wakeling-sociomantic replied:
Yup, will do so.  Just wanted to get the design itself tied down/approved first.

How feasible is it to really _test_ something like this?  I don't have much experience of swarm/proto unittesting and integration tests.

mihails-strasuns-sociomantic replied:
Still missing usage example

gavin-norman-sociomantic commented:
Generally, this seems like a nice, convenient, little helper :+1:

gavin-norman-sociomantic commented:
The commit message can say "fixes #126", btw.

Oh: release notes! :police_car:

joseph-wakeling-sociomantic commented:
Updated to use `TaskEvent` and to rename the `dg` delegates.  I'll follow up with release notes and a unittest usage example.

joseph-wakeling-sociomantic commented:
... oh, and to remove the `epoll` constructor parameter and use `theScheduler.epoll` instead.

joseph-wakeling-sociomantic commented:
Release notes, usage examples, and issue cross-refs to follow ;-)

mihails-strasuns-sociomantic reviewed (commented):
Also red ;)

joseph-wakeling-sociomantic commented:
Turns out we have a small problem with a usage example: `dhtproto`'s ocean dependency predates the introduction of `Scheduler.awaitOrTimeout` ;-)

joseph-wakeling-sociomantic commented:
Usage example pushed.  I can see two ways to deal with this:

* bork the usage example to only use `await` (and fix it in dhtproto 14.x.x)

  * maybe noting the possibility in comments

* use ocean versioning to select whether or not to use `awaitOrTimeout`

Any preferences?

mihails-strasuns-sociomantic commented:
I think first option is better(noting alternative in comments) - version blocks in usage example feel real confusing.

joseph-wakeling-sociomantic commented:
OK. I will just use `await` and note in a comment that `awaitOrHandshake` can be used (where available) to implement a partial handshake.  And then once this is merged into 14.x.x I'll make a follow-up PR to improve things there.

joseph-wakeling-sociomantic commented:
Updated with to more to-squash patches:

* fixed usage example to work with minimum supported ocean version

* added release notes, with a more up to date usage example

joseph-wakeling-sociomantic commented:
@mihails-strasuns-sociomantic if this looks OK to you, let me know and I'll squash it all down for final review ;-)

joseph-wakeling-sociomantic commented:
... ping? :-P

nemanja-boric-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,241 @@
+
+    ***************************************************************************/
+
+    public this (DhtClient dht, ulong wait_time)
```
Nitpick: I like `retry_delay_seconds` (as used in the example) more than `wait_time`.

joseph-wakeling-sociomantic replied:
Sure, can do.  I was copying `wait_time` from `RetryHandshake`, but I agree `retry_delay_seconds` is nicer.

joseph-wakeling-sociomantic commented:
`fixup` to rename `wait_time` => `retry_delay_seconds`.  Anything else that needs addressing, before I squash?

nemanja-boric-sociomantic commented:
LGTM. @mihails-strasuns-sociomantic ?

mihails-strasuns-sociomantic reviewed (approved):
Very nice!

joseph-wakeling-sociomantic commented:
Squashed and rebased on latest 13.x.x, and corrected a few small errors in the release-notes usage example:

```diff
diff --git a/relnotes/handshake.feature.md b/relnotes/handshake.feature.md
index 18aef68..610950a 100644
--- a/relnotes/handshake.feature.md
+++ b/relnotes/handshake.feature.md
@@ -8,10 +8,10 @@ applications to support a partial handshake, e.g.:
 
 ```D
 auto retry_delay_seconds = 3;
-handshake = new DhtHandshake(dht_client, retry_delay_seconds);
+auto handshake = new DhtHandshake(dht_client, retry_delay_seconds);
 
 // block on at least one node connecting
-theScheduler.await(this.handshake.oneNodeConnected());
+theScheduler.await(handshake.oneNodeConnected());
 Stdout.formatln("At least one node is now connected!");
 
 // wait until either all nodes have connected, or 60 seconds
@@ -21,7 +21,7 @@ auto timeout_microsec = 60_000_000;
 
 auto handshake_timeout =   // true if timeout is reached
     theScheduler.awaitOrTimeout(
-        this.handshake.allNodesConnected(),
+        handshake.allNodesConnected(),
         timeout_microsec);
 
 if (handshake_timeout)
```

joseph-wakeling-sociomantic commented:
BTW it would be great if this could be merged early from 13.x.x into 14.x.x, because I'd like to update the imports (`Scheduler` => `IScheduler`) and usage example in that release.

joseph-wakeling-sociomantic commented:
Oh wait!  Need one last commit-message update ...

joseph-wakeling-sociomantic commented:
Updated commit message and PR description with cross-ref to https://github.com/sociomantic-tsunami/dhtproto/issues/126.

joseph-wakeling-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,241 @@
+
+    ***************************************************************************/
+
+    public this (DhtClient dht, ulong retry_delay_seconds)
```
Minor note: `RetryHandshake`'s `wait_time` parameter is `size_t`.  Perhaps we should use `uint` here to guarantee it is compatible?  (Yes, we only support 64-bit, but this is low-hanging fruit to avoid incompatibility with 32-bit.)

mihails-strasuns replied:
We should also use `size_t`.

joseph-wakeling-sociomantic replied:
Why?  It's a weird type to use for a value in seconds.

mihails-strasuns replied:
To match upstream definition and avoid 32-bit incompatibility.

joseph-wakeling-sociomantic replied:
`uint` will be compatible with the `size_t` parameter of `RetryHandshake`, as will `int` with a `verify(retry_delay_seconds > 0)` under the hood (@nemanja-boric-sociomantic expressed a preference for a signed value).

nemanja-boric-sociomantic commented:
> BTW it would be great if this could be merged early from 13.x.x into 14.x.x, because I'd like to update the imports (Scheduler => IScheduler) and usage example in that release.

I'll leave this decision to @mihails-strasuns-sociomantic, as he was against merging non-tags between the released branches .

mihails-strasuns commented:
I am fine with merging non-tagged version branches, I was objecting merges in/of major feature branches.

joseph-wakeling-sociomantic commented:
OK, in the absence of consensus on any other solution, I've updated the constructor parameter to use `size_t` to match `RetryHandshake`:

```diff
diff --git a/src/dhtproto/client/legacy/internal/helper/Handshake.d b/src/dhtproto/client/legacy/internal/helper/Handshake.d
index c924491..e07c900 100644
--- a/src/dhtproto/client/legacy/internal/helper/Handshake.d
+++ b/src/dhtproto/client/legacy/internal/helper/Handshake.d
@@ -131,7 +131,7 @@ public class DhtHandshake
 
     ***************************************************************************/
 
-    public this (DhtClient dht, ulong retry_delay_seconds)
+    public this (DhtClient dht, size_t retry_delay_seconds)
     in
     {
         assert(dht !is null);
```

joseph-wakeling-sociomantic commented:
... good to go? :-)

mihails-strasuns commented:
As soon as @gavin-norman-sociomantic is back from vacation ;)

joseph-wakeling-sociomantic commented:
... _Et exspecto resurrectionem mortuorum..._ ;-)

gavin-norman-sociomantic commented on src/dhtproto/client/legacy/internal/helper/Handshake.d:
```diff
@@ -0,0 +1,241 @@
+    public this (DhtClient dht, size_t retry_delay_seconds)
+    in
+    {
+        assert(dht !is null);
```
Could you check here that the scheduler has really been initialised? (i.e. that the epoll instance to be used is valid.)

joseph-wakeling-sociomantic replied:
Hmmm, is there a mechanism for this in the ocean version dhtproto 13.x.x needs to support?

More recent ocean versions define an `isSchedulerUsed` function that can be used to check that, but that will only be available from dhtproto 14.x.x onwards.

Note that `theScheduler` has an out-contract that verifies that the underlying `Scheduler` instance has been initialized (i.e. it can't return `null`, and therefore in turn `initScheduler` must have been called).

gavin-norman-sociomantic replied:
Oh that's fine then :+1:.

joseph-wakeling-sociomantic replied:
Cool!  Let me know once it's merged into 14.x.x: it may be worth tweaking it to use `IScheduler`, and we _can_ use `isSchedulerUsed()` as an extra check if wanted.

gavin-norman-sociomantic reviewed (commented):
Apart from the one comment, looks great!
